docs: blog post on Avoiding `any`s with Linting and TypeScript (#10362)
typescript-eslint/typescript-eslint: packages/website/blog/2025-01-13-avoiding-anys.mdx

[`@typescript-eslint/use-unknown-in-catch-callback-variable`](/rules/use-unknown-in-catch-callback-variable) enforces always using the unknown type for the parameter of a Promise rejection callback.

For example, given the following code, TypeScript would not report a type error, but the lint rule would report:

```ts
function rejectWith(value: string) {
  return Promise.reject(value);
}

rejectWith('Nandor').catch(error => {
  //                       ~~~~~
  // eslint(@typescript-eslint/use-unknown-in-catch-callback-variable):
  // Prefer the safe `: unknown` for a `catch` callback variable.

  console.log(error.message); // Should be a type error, but isn't
});
```

Had the `error` been given a `: unknown` type, TypeScript would be able to report on `error.message` as being unsafe.

## Banning Usage of Unsafe Types

Once an `any` type exists in code, it is "infectious": it can turn the types of values based on it into more `any`s.

typescript-eslint includes a collection of rules that flag code patterns which spread or are generally made less safe due to an `any`.

> Each of the following rules reports on a specific use of `any`.

Author: Yeah agreed. I'll further condense this by only including one example out of the list.

  }
  ```

* [`@typescript-eslint/no-unsafe-call`](/rules/no-unsafe-call) reports on calling an `any` typed value like a function:

  ```ts
  const callbacks = {
    name: (value: string) => console.log(value);
  };

  function setNameFromData(raw: string) {
    const callback = JSON.parse(raw);

    callback();
    // ~~~~~~~
    // eslint(@typescript-eslint/no-unsafe-call):
    // Unsafe assignment of an `any` value.
  }
  ```

* [`@typescript-eslint/no-unsafe-member-access`](/rules/no-unsafe-member-access) reports on accessing members of any value typed as `any`:

  ```ts
  function setNameOnData(raw: string) {
    const data = JSON.parse(raw);

    data.name = 'Colin';
    //   ~~~~
    // eslint(@typescript-eslint/no-unsafe-member-access):
    // Unsafe member access .name on an `any` value.
  }
  ```

* [`@typescript-eslint/no-unsafe-return`](/rules/no-unsafe-return) reports on returning a value typed as `any` from a function:

  ```ts
  function returnData(raw: string) {
    return JSON.parse(raw);
    // ~~~~~~~~~~~~~~~~~~~~
    // eslint(@typescript-eslint/no-unsafe-return):
    // Unsafe return of a value of type `any`.
  }
  ```

> Put together, the `@typescript-eslint/no-unusafe-*` rules provide a comprehensive suite of checks that can flag most accidental uses of `any` across a codebase.

Author: Best I got is mentioning _"around `any` safety"_

Those inline comment directives cause TypeScript to ignore type errors on a line.
Disabling TypeScript on a per-line basis can sometimes be necessary in rare cases, but is always unsafe and almost never the right choice.

:::tip
`// @ts-expect-error` and `// @ts-ignore` are almost the same, except `// @ts-expect-error` will produce a new type error if its line doesn't already have one.
It's generally preferable to use `// @ts-expect-error` over `// @ts-ignore`.
:::

[`@typescript-eslint/ban-comment`](/rules/ban-ts-comment) can be used to report on cases where developers use TypeScript comment directives.
By default, the rule:

- Always reports on `@ts-ignore` and `@ts-nocheck` comment directives
- Reports on `@ts-expect-error` directives that don't have an explanatory comment description
- Requires explanatory comment descriptions contain at least 3 characters

For example, suppose `"@example/package"` exports a `process` function that should take in a `string` but whose types incorrectly indicate take in a `number`.
A `// @ts-expect-error` comment could be used to tell TypeScript to ignore the line:

<Tabs>
<TabItem value="❌ Incorrect">

```ts

> import { process } from '@example/package';

Reviewer: Consider avoiding red herring collision with built-in [`globalThis.process`](https://nodejs.org/docs/v22.13.0/api/globals.html#process)
```suggestion
import { processString } from '@example/package';
```

(here and the other tab)

// @ts-ignore
process('New York City');
```

</TabItem>
<TabItem value="✅ Correct">

```ts
import { process } from '@example/package';

// @ts-expect-error -- Pending updating the process types. See GH-1234.
process('New York City');
```

</TabItem>
</Tabs>

Once the types for `@example/package` are fixed to no longer produce a type error, the comment directive will itself produce a type error asking to delete itself.

### ESLint Comments Plugin

ESLint includes its own inline comment directives separate from TypeScript's.
`// eslint-disable`, `// eslint-disable-next-line`, and other inline ESLint config comments can suppress lint rules — including those mentioned in this post.

> Suppressing ESLint rules can also occasionally be necessary in cases where the rule and/or TypeScript's type system have misunderstand your code.

Reviewer: nit - honestly "occasionally" feels a little strong here. It's not that uncommon that we tell users to use eslint-disable for rule limitations and such.

```suggestion
Suppressing ESLint rules can also be necessary in cases where the rule and/or TypeScript's type system have misunderstand your code.
```

Reviewer: The subsequent "absolutely must" also a little too strong IMO

If you absolutely must use ESLint comment directives, we recommend utilizing [`@eslint-community/eslint-plugin-eslint-comments`](https://eslint-community.github.io/eslint-plugin-eslint-comments).
The plugin's [recommended preset](https://eslint-community.github.io/eslint-plugin-eslint-comments/#%F0%9F%93%96-usage) enables rules that enforce best practices around ESLint comment directives, including:

- [`@eslint-community/eslint-comments/disable-enable-pair`](https://eslint-community.github.io/eslint-plugin-eslint-comments/rules/disable-enable-pair.html): to prevent accidentally disabling rules for the rest of a file, rather than on specific lines
- [`@eslint-community/eslint-comments/no-unlimited-disable`](https://eslint-community.github.io/eslint-plugin-eslint-comments/rules/no-unlimited-disable.html): to prevent accidentally disabling all ESLint rules, rather than specific ones
- [`@eslint-community/eslint-comments/require-description`](https://eslint-community.github.io/eslint-plugin-eslint-comments/rules/require-description.html): to require comments explaining why directives are necessary

> For example, suppose the type of an imported `process` is temporarily `any` pending some soon-to-be-completed task.

Reviewer: Same feedback about the use of `process` here. Coupled with the rest of the sentence, the overall sentence right now reads to me as though it's describing runtime behavior not static analysis types.

```suggestion
For example, suppose the type of an imported variable is temporarily `any` pending some soon-to-be-completed refactoring.
```
